File: server/src/main/java/cn/iinti/atom/service/base/alert/events/MetricMonitorConfig.java
package cn.iinti.atom.service.base.alert.events;

import cn.iinti.atom.service.base.alert.MetricMonitorHandle;
import cn.iinti.atom.service.base.metric.MetricEnums;
import cn.iinti.atom.service.base.metric.MetricVo;
import cn.iinti.atom.service.base.metric.mql.MQL;
import groovy.lang.Closure;
import groovy.lang.DelegatesTo;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.apache.commons.lang3.StringUtils;
import org.springframework.scheduling.support.CronExpression;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;


public class MetricMonitorConfig {
    @Getter
    private String _id = UUID.randomUUID().toString();
    private CronExpression _cron;
    private MQL _mql;
    private LocalDateTime _start;
    private LocalDateTime _end;
    private MetricEnums.MetricAccuracy _accuracy = MetricEnums.MetricAccuracy.hours;

    private Closure<?> callback;

    public void id(String id) {
        this._id = id;
    }

    public String getId() {
        return this._id;
    }

    public void cron(String con) {
        this._cron = CronExpression.parse(con);
    }

    public void mql(String mql) {
        this._mql = MQL.compile(mql);
    }

    public void start(LocalDateTime start) {
        this._start = start;
    }

    public void end(LocalDateTime end) {
        this._end = end;
    }

    public void accuracy(MetricEnums.MetricAccuracy accuracy) {
        this._accuracy = accuracy;
    }

    @AllArgsConstructor
    public static class MetricData {
        private final Map<String, List<MetricVo>> metric;

        public double getMetricValue(String key, String... tags) {
            Double value = getMetric(key, tags);
            return value != null ? value : 0.0;
        }

        public Double getMetric(String key, String... tags) {
            List<MetricVo> metricVos = metric.get(key);
            if (metricVos == null || metricVos.isEmpty()) {
                return null;
            }
            if (metricVos.size() == 1) {
                return metricVos.get(0).getValue();
            }
            // filter by tags
            Map<String, String> tagMap = new HashMap<>();
            for (int i = 0; i < tags.length; i += 2) {
                tagMap.put(tags[i], tags[i + 1]);
            }
            for (MetricVo metricVo : metricVos) {
                if (isTagEquals(metricVo.getTags(), tagMap)) {
                    return metricVo.getValue();
                }
            }
            // warning
            return null;
        }

        private static boolean isTagEquals(Map<String, String> tags1, Map<String, String> tags2) {
            if (tags1.size() != tags2.size()) {
                return false;
            }
            for (Map.Entry<String, String> entry : tags1.entrySet()) {
                String k = entry.getKey();
                String v = entry.getValue();
                if (!StringUtils.equals(v, tags2.get(k))) {
                    return false;
                }
            }
            return true;
        }
    }

    public void valid() {
        if (_mql == null) {
            throw new IllegalStateException("mql not set");
        }
        if (_cron == null) {
            throw new IllegalStateException("cron not set");
        }
        if (callback == null) {
            throw new IllegalStateException("onMetric callback is null");
        }
    }

    public void onMetric(@DelegatesTo(MetricData.class) Closure<?> closure) {
        callback = closure;
    }

    public void fillMeta(MetricMonitorHandle metricMonitorHandle) {
        metricMonitorHandle.setMql(_mql);
        metricMonitorHandle.setCron(_cron);
        metricMonitorHandle.setCallback(callback);
        metricMonitorHandle.setStart(_start);
        metricMonitorHandle.setEnd(_end);
        metricMonitorHandle.setAccuracy(_accuracy);
    }
}

File: server/src/main/kotlin/cn/iinti/atom/service/base/alert/events/MetricMonitorConfig.kt
package cn.iinti.atom.service.base.alert.events

import cn.iinti.atom.service.base.alert.MetricMonitorHandle
import cn.iinti.atom.service.base.metric.MetricEnums
import cn.iinti.atom.service.base.metric.MetricVo
import cn.iinti.atom.service.base.metric.mql.MQL
import groovy.lang.Closure
import groovy.lang.DelegatesTo
import org.apache.commons.lang3.StringUtils
import org.springframework.scheduling.support.CronExpression
import java.time.LocalDateTime
import java.util.*

class MetricMonitorConfig {
    private var _id: String = UUID.randomUUID().toString()
    private var _cron: CronExpression? = null
    private var _mql: MQL? = null
    private var _start: LocalDateTime? = null
    private var _end: LocalDateTime? = null
    private var _accuracy: MetricEnums.MetricAccuracy = MetricEnums.MetricAccuracy.hours

    private var callback: Closure<*>? = null

    fun id(id: String) {
        this._id = id
    }

    fun getId(): String {
        return this._id
    }

    fun cron(con: String) {
        this._cron = CronExpression.parse(con)
    }

    fun mql(mql: String) {
        this._mql = MQL.compile(mql)
    }

    fun start(start: LocalDateTime) {
        this._start = start
    }

    fun end(end: LocalDateTime) {
        this._end = end
    }

    fun accuracy(accuracy: MetricEnums.MetricAccuracy) {
        this._accuracy = accuracy
    }

    class MetricData(private val metric: Map<String, List<MetricVo>>) {
        fun getMetricValue(key: String, vararg tags: String): Double {
            val value = getMetric(key, *tags)
            return value ?: 0.0
        }

        fun getMetric(key: String, vararg tags: String): Double? {
            val metricVos = metric[key]
            if (metricVos == null || metricVos.isEmpty()) {
                return null
            }
            if (metricVos.size == 1) {
                return metricVos[0].value
            }
            // filter by tags
            val tagMap = HashMap<String, String>()
            for (i in tags.indices step 2) {
                if (i + 1 < tags.size) {
                    tagMap[tags[i]] = tags[i + 1]
                }
            }
            for (metricVo in metricVos) {
                if (isTagEquals(metricVo.tags, tagMap)) {
                    return metricVo.value
                }
            }
            // warning
            return null
        }

        companion object {
            fun isTagEquals(tags1: Map<String, String>, tags2: Map<String, String>): Boolean {
                if (tags1.size != tags2.size) {
                    return false
                }
                for ((k, v) in tags1) {
                    if (!StringUtils.equals(v, tags2[k])) {
                        return false
                    }
                }
                return true
            }
        }
    }

    fun valid() {
        if (_mql == null) {
            throw IllegalStateException("mql not set")
        }
        if (_cron == null) {
            throw IllegalStateException("cron not set")
        }
        if (callback == null) {
            throw IllegalStateException("onMetric callback is null")
        }
    }

    fun onMetric(@DelegatesTo(MetricData::class) closure: Closure<*>) {
        callback = closure
    }

    fun fillMeta(metricMonitorHandle: MetricMonitorHandle) {
        metricMonitorHandle.setMql(_mql)
        metricMonitorHandle.setCron(_cron)
        metricMonitorHandle.setCallback(callback)
        metricMonitorHandle.setStart(_start)
        metricMonitorHandle.setEnd(_end)
        metricMonitorHandle.setAccuracy(_accuracy)
    }
}